2023/09/testing-user-defined-functions-with-pgtap.md

```diff
@@ -0,0 +1,462 @@
+database for MySQL, called Pagila. We’ll construct two functions and go over a single test case. Using basic Postgres tools, I’ll walk through a few methods of calling and testing
+our functions. Finally, we’ll go over a more concise way of testing all cases for said functions. Feel free to work alongside this article and hopefully leave with a comfortable 
+idea of how to construct unit test files for your database.
+
+### Functions:
+
+*Note: If performing this first example, remember to have the pgTAP extension in your schema and perform it in a transaction (I’ve omitted `BEGIN` and `ROLLBACK`). Also use `SELECT
+plan(<number of tests you’ll run>)` when running the pgTAP functions.*
+
+Starting off, let's create a basic addition function that returns an INTEGER value:
+
+```postgresql
+CREATE OR REPLACE FUNCTION add_numbers(
+    num1 INTEGER,
+    num2 INTEGER
+) 
+RETURNS INTEGER 
+LANGUAGE plpgsql AS $$
+BEGIN
+    RETURN num1 + num2;
+END;
+$$;
+```
+To test this user-defined function, a few of the most relevant functions in pgTAP we can use are:
+
+#### `has_function()`
+
+Variations: 
+```postgresql
+SELECT has_function( :schema, :function, :args, :description );
+SELECT has_function( :schema, :function, :args );
+SELECT has_function( :schema, :function, :description );
+SELECT has_function( :schema, :function );
+SELECT has_function( :function, :args, :description );
+SELECT has_function( :function, :args );
+SELECT has_function( :function, :description );
+SELECT has_function( :function );
+```
+
+Example:
+```postgresql
+SELECT has_function('your_schema', 'add_numbers', ARRAY['INTEGER','INTEGER']);   -- default description used
+```
+Results:
+```postgresql 
+ok 1 - Function add_numbers(INTEGER, INTEGER) should exist
+```
+
+Parameters
+```
+:schema
+    Name of a schema in which to find the function.
+:function
+    Name of a function.
+:args
+    Array of data types of the function arguments.
+:description
+    A short description of the test.
+```
+
+`has_function()` is useful in checking the existence of a specific function. If you want to parse an entire schema, `functions_are()` is your command with pretty straightforward parameters.
+
+#### `functions_are()`
+Variations: 
+```postgresql
+SELECT functions_are( :schema, :functions, :description );
+SELECT functions_are( :schema, :functions );
+SELECT functions_are( :functions, :description );
+SELECT functions_are( :functions );
+```
+Example:
+```postgresql
+SELECT functions_are('your_schema', ARRAY['add_numbers', <All other functions need to be listed as well>]);   -- default description used
+```
+Parameters
+```
+:schema
+    Name of a schema in which to find functions.
+:functions
+    An array of function names.
+:description
+    A short description of the test.
+```
+#### `ok()` and `is()`
+
+`ok()` and `is()` have similar a function with slightly different parameters:
```

Comment:
Maybe this should say 
`ok()` and `is()` have similar functionality with slightly different parameters